Open amp refactoring and add more platforms support

karl-zh opened the pull request:
openAMP is a multiple core IPC which supported in Zephyr. This PR is going to make it supported for more multiple core platforms rather than for lpc54 only.

Currently added the arm platform AN521 and Musca A, for further plan is to supporting Musca B1 when #16729 get merged.

Commit:
samples: openamp: Remove exclusive compile of openAMP

Add configuable shared memory address for openAMP samples. There is a
plan to add more platforms supported for openAMP in zephyr.

Each platform can specify the shared memory address and device by
device tree and add it's support in openAMP samples.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

Commit:
samples: openamp: Add support for mps2_an521

AN521 is a dual core FPGA on MPS2+ with both cores are CM33. Add openAMP
to support on it.

Core 0 is primary core, it runs as master, core 1 is remote, it runs
as slave.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

Commit:
samples: openamp: Add support for Musca A

Musca A is a dual core SoC with both cores are CM33. Add
openAMP to support on it.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

Commit:
samples: openamp: Keep document up to date

Update serial output for both master and remote.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

galak commented:
@arnop2 can you take a look at this PR.  I'm not that familiar with how we should be conveying the shared memory region in DTS.

arnop2 commented on boards/arm/mps2_an521/mps2_an521_nonsecure.dts:
```diff
@@ -45,6 +45,11 @@
+		compatible = "mmio-sram";
+		reg = <0x28180000 0x8000>;
+	};
+
```
Very interesting implementation based on the memory declaration in the DTS. I hesitated to do same in my pull request (https://github.com/zephyrproject-rtos/zephyr/pull/16985).
A drawback in this implementation is that the memory region is project dependent. It is needed only for OpenAMP, but declared for a board, so for all projects.
For instance, this result here in an overlap between the memory declared for the zephyr firmware and this shared memory. not safe...
An other alternative would be  to limit the zephyr SRAM, but impact is for all projects...

So a point to solve here (IMHO) is to make this shared region optional ( project dependent) to allow to use it for some other purposes. But also to configure the rest of the ram according to this shared memory.

To get around the issue, In my pull request, i have implemented the memory pool  as a linker memory region, under CONFIG_OPENAMP. But I'm not satisfied with my implementation. If my shared memory is optional, This does not prevent overlap with the zephyr SRAM. 

I have no real solution  to propose:), i can just try to clarify the equation:
- There is a dependency on project: use of a CONFIG_OPENAMP?
- There is a dependency on the platform ( memory region definition + Zephyr RAM resizing): DTS overlay, linker with some CONFIG values, ...?

arnop2 commented on boards/arm/mps2_an521/mps2_an521.dts:
```diff
@@ -45,6 +45,11 @@
 
+	sramx: memory@38180000 {
+		compatible = "mmio-sram";
+		reg = <0x38180000 0x8000>;
```
strange address regarding other dts files of you pull request...Is it a specific normal that this memory is  not include in sram0?

karl-zh replied:
Yes, this seems strange, not included in the normal sram and also for all projects rather than openAMP itself.

Then, how about define the shared memory address as a macro in soc.h with the control of CONFIG_OPENAMP defined?

If using CONFIG_SHM_ADDR_xxx, that may require more prj.conf files for platfrom specific I think.

@galak

arnop2 replied:
I don't know, and probably I'm not expert enough in zephyr to determine the best solution.....
For a project integrating OpenAMP, I assume that the good practice is to define the shared memory in DTS and to resize the zephyr RAM accordingly. 
I don't know if it would be possible to create a dts overlay associated with the example project...? I couldn't find any examples.

arnop2 replied:
Sorry I'm Linux formatted, so i looked for .dtso files... overlay is possible for sample ( but maybe you already know this). There are a lot of .overlay files in samples. so you could use similar implementation. 
For instance ( code not tested) you could create a  mps2_an521_nonsecure.overlay:
&sram0: {
		reg = <0x28100000 0xF80000>;
	};

/ {
	sramx: memory@0x281F8000 {
		compatible = "mmio-sram";
		reg = <0x281F8000 0x8000>;
	};
};

This limits the RAM used by the linker script, so ensure coherency in memory allocations, and allow to declare the memory assigned to the OpenAMP  in an elegant way. So it seems like a good compromise (IMHO).

karl-zh replied:
Hi Arnaud,

Thanks for your suggestion of overlay. I updated it now.

dbkinder commented on samples/subsys/ipc/openamp/README.rst:
```diff
@@ -27,17 +43,31 @@ following settings:
 
 Reset the board and the following message will appear on the corresponding
-serial port:
+serial port, one is master another is slave:
```
should ``slave`` be ``remote``, since that's what shown in the sample output below?

karl-zh replied:
Hi David
Fixed now. Thanks for your review.

dbkinder reviewed (changes requested):
one doc question, otherwise looks good.

arnop2 commented on soc/arm/arm/mps2/dts_fixup.h:
```diff
@@ -36,6 +36,11 @@
+#define DT_VDEV_SIZE				DT_MMIO_SRAM_28180000_SIZE
+
+#define DT_IPM_DEV				DT_ARM_MHU_40003000_LABEL
+
```
With overlays in sample you need to use a chosen{} defined in the overlay instead
(https://docs.zephyrproject.org/latest/guides/dts/index.html?highlight=chosen#aliases-and-chosen-nodes)

arnop2 replied:
@karl-zh  
I updated my PR (https://github.com/zephyrproject-rtos/zephyr/pull/16985). I implemented the DT overlay in the sample with the use of chosen{} to have a generic name for the shared memory pool.  To do similar implementation, you have to get this commit : https://github.com/zephyrproject-rtos/zephyr/pull/16985/commits/19309c7a5608b6c2b6f06dab9fceb9edba8a9acf. Then you have to create the overlay using the template <board>.overlay( including the memory and the chosen declarations ). The overlay will be automatically applied on top of and DT_IPC_SHM_BASE_ADDRESS + DT_IPC_SHM_BASE_SIZE defines will be generated.

@galak could you confirm if you are in line with this DT overlay approach

dbkinder reviewed (approved):
+1 for doc changes, thanks.

karl-zh commented on samples/subsys/ipc/openamp/sse_200.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_MAIN_STACK_SIZE=2048
+CONFIG_HEAP_MEM_POOL_SIZE=4096
+CONFIG_OPENAMP=y
+CONFIG_IPM_MHU=y
```
Okay, the IPM MHU could be in soc or board level I think.
It seems still have some specific for lpcx, like:
`CONFIG_IPM_MCUX=y
CONFIG_SLAVE_CORE_MCUX=y
CONFIG_SLAVE_IMAGE_MCUX="${ZEPHYR_BINARY_DIR}/../openamp_remote-prefix/src/openamp_remote-build/zephyr/zephyr.bin"`

sse_200.conf is for AN521, Musca A and Musca B for common.

arnop2 replied:
The stm32H7 Soc should be able to reuse it also (need TBC).
The idea here would be  to have something similar to https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/bluetooth/hci_uart . A generic conf file that can be used by default or used as template to adapt to a specific platform need.
This is of course  a proposal and not a request...

Commit:
IPM: Add MHU as default IPM for sse-200 subsystem

Set sse-200 subsystem using MHU (Message Handling Unit) as default.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

arnopo reviewed (changes requested):
Just 2 minor remarks remaining, then  should be ok for me

karl-zh commented on samples/subsys/ipc/openamp/CMakeLists.txt:
```diff
@@ -8,6 +8,9 @@ cmake_minimum_required(VERSION 3.13.1)
   set(DTC_OVERLAY_FILE ${CMAKE_CURRENT_SOURCE_DIR}/overlay/nxp_lpc54xxx.overlay)
   set(BOARD_REMOTE "lpcxpresso54114_m0")
+elseif("${BOARD}" STREQUAL "mps2_an521")
+  set(DTC_OVERLAY_FILE ${CMAKE_CURRENT_SOURCE_DIR}/overlay/${BOARD}.overlay)
```
Fine, I was thinking to put all overlay files into a folder, since there will be more platforms supported. Will update later.

arnop2 replied:
Reading my comment, i not sure that my comment is clear... so in doubt:
When i say the "overlay file is automatically set", i mean DTC_OVERLAY_FILE environment variable is automatically set.

karl-zh replied:
Yes. Thanks, I have changed this to using automatically.

karl-zh commented on samples/subsys/ipc/openamp/common.h:
```diff
@@ -1,23 +1,28 @@
+
+#define SHM_START_ADDR		(VDEV_START_ADDR + VDEV_STATUS_SIZE)
+#define SHM_SIZE		(VDEV_SIZE - VDEV_STATUS_SIZE)
 #define SHM_DEVICE_NAME		"sramx.shm"
```
Hi Arnaud:
Okay, this is derived from the original demo, it is a shared memory based on ram. So I think it is make sense to continue use "sramx.shm". Do you have some ideas on this?
Thanks

arnop2 replied:
Just the "sramx" which can make no sense depending on platform. 
 " sh_mem" could be more generic, but feel free to choice a better name :)

erwango commented on scripts/dts/extract/globals.py:
```diff
@@ -25,7 +25,8 @@
     'zephyr,ccm'   : 'DT_CCM',
-    'zephyr,dtcm'  : 'DT_DTCM'
+    'zephyr,dtcm'  : 'DT_DTCM',
+    'zephyr,ipc_shm' : 'DT_IPC_SHM'
```
Expect conflicts with #17660

erwango replied:
#17660 is now merged, you'll have to report this patch in `scripts/dts/gen_defines.py`

arnop2 replied:
related commit updated in https://github.com/zephyrproject-rtos/zephyr/pull/16985

karl-zh replied:
Updated.

arnopo reviewed (approved):
Just a minor request to change SHM_DEVICE_NAME definition + a rebase to fix my patch commit messag ( apology for this).
Then LGTM, thanks.

karl-zh commented:
> Just a minor request to change SHM_DEVICE_NAME definition + a rebase to fix my patch commit messag ( apology for this).
> Then LGTM, thanks.

Thanks. Let us open for more ideas, try to make everyone happy on this. 👍

Commit:
samples: openamp: Add support for Musca B1

Musca B1 is a dual core SoC with both cores are CM33. Add
openAMP to support on it.

Signed-off-by: Karl Zhang <karl.zhang@linaro.org>

galak committed:
samples: openamp: prep for additional board support

Move board specific conf info boards/<BOARD>.conf to support other
board/SoCs with this sample.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented on samples/subsys/ipc/openamp/remote/src/main.c:
```diff
@@ -73,9 +73,17 @@ static u32_t virtio_get_features(struct virtio_device *vdev)
+	defined(CONFIG_SOC_V2M_MUSCA_B1)
+	u32_t current_core = sse_200_platform_get_cpu_id();
+
+	ipm_send(ipm_handle, 0, current_core ? 0 : 1, 0, 1);
```
Why does the data here mater?

karl-zh replied:
CPU0 send notification to CPU1, vise vesa.

karl-zh replied:
The dummy data is not necessary here for MHU based hardware. But it needs to distinguish the current core No. for notifications.

galak commented:
Ok, I tried this on musca-a1 and qemu-mps2_an521 and it doesn't seem to work.  Also tried the `samples/subsys/ipc/ipm_mhu_dual_core` on those platforms and they didn't seem to pass.  Something to debug this week.

karl-zh commented:
> Ok, I tried this on musca-a1 and qemu-mps2_an521 and it doesn't seem to work. Also tried the `samples/subsys/ipc/ipm_mhu_dual_core` on those platforms and they didn't seem to pass. Something to debug this week.

Okay, I also need some refresh of memory to this context. BTW, do you have some fail context? I just have a test pass on my side with the previous commit I had on AN521, here is the command line for you reference:

**AN521:** 
~/zephyr/samples/subsys/ipc/openamp/build$ rm -rf *; rm -rf .* ; cmake -GNinja -DBOARD=mps2_an521 -DCONF_FILE=prj.conf  ..; ninja -v; cp ./openamp_remote-prefix/src/openamp_remote-build/zephyr/zephyr.bin ../zephyr_ns.bin ; cp zephyr/zephyr.bin ../zephyr.bin

Config file:
TITLE: Versatile Express Images Configuration File

[IMAGES]
TOTALIMAGES: 2                     ;Number of Images (Max: 32)

IMAGE0ADDRESS: 0x10000000          ;Please select the required executable program
IMAGE0FILE: \SW\zephyr.bin  ; - selftest

IMAGE1ADDRESS: 0x100000
IMAGE1FILE: \SW\zephyr_ns.bin

galak commented:
> Okay, I also need some refresh of memory to this context. Maybe some rebase conflicts or other changes need rebase. BTW, do you have some fail details? I just have a test pass on my side with the previous commit I had on AN521, here is the command line for you reference:
> 
> **AN521:**
> ~/zephyr/samples/subsys/ipc/openamp/build$ cmake -GNinja -DBOARD=mps2_an521 -DCONF_FILE=prj.conf ..; ninja -v; cp ./openamp_remote-prefix/src/openamp_remote-build/zephyr/zephyr.bin ../zephyr_ns.bin ; cp zephyr/zephyr.bin ../zephyr.bin
> 
> Config file:
> TITLE: Versatile Express Images Configuration File
> 
> [IMAGES]
> TOTALIMAGES: 2 ;Number of Images (Max: 32)
> 
> IMAGE0ADDRESS: 0x10000000 ;Please select the required executable program
> IMAGE0FILE: \SW\zephyr.bin ; - selftest
> 
> IMAGE1ADDRESS: 0x100000
> IMAGE1FILE: \SW\zephyr_ns.bin

Hmm, need to parse this in more detail.  However is it possible to run the test w/o SECURE/NON-SECURE?

karl-zh commented:
> > Okay, I also need some refresh of memory to this context. Maybe some rebase conflicts or other changes need rebase. BTW, do you have some fail details? I just have a test pass on my side with the previous commit I had on AN521, here is the command line for you reference:
> > **AN521:**
> > ~/zephyr/samples/subsys/ipc/openamp/build$ cmake -GNinja -DBOARD=mps2_an521 -DCONF_FILE=prj.conf ..; ninja -v; cp ./openamp_remote-prefix/src/openamp_remote-build/zephyr/zephyr.bin ../zephyr_ns.bin ; cp zephyr/zephyr.bin ../zephyr.bin
> > Config file:
> > TITLE: Versatile Express Images Configuration File
> > [IMAGES]
> > TOTALIMAGES: 2 ;Number of Images (Max: 32)
> > IMAGE0ADDRESS: 0x10000000 ;Please select the required executable program
> > IMAGE0FILE: \SW\zephyr.bin ; - selftest
> > IMAGE1ADDRESS: 0x100000
> > IMAGE1FILE: \SW\zephyr_ns.bin
> 
> Hmm, need to parse this in more detail. However is it possible to run the test w/o SECURE/NON-SECURE?

I might using a wrong binary name here. They are following the start work for tfm integrating with Zpehyr.
The two binaries here are for different CPUs, \SW\zephyr.bin is for CPU0 and \SW\zephyr_ns.bin is for CPU1,  the name is for configurations file only(images.txt).

jhedberg commented:
This PR seems to require rebasing since there are conflicts. Is something still missing before it can be merged (e.g. approvals or some additional code changes)?

karl-zh commented:
> This PR seems to require rebasing since there are conflicts. Is something still missing before it can be merged (e.g. approvals or some additional code changes)?

Hi Johan

I think this PR needs approval before the merge.

jhedberg commented:
> I think this PR needs approval before the merge.

@karl-zh while waiting for approval, could you rebase and fix the conflict with `samples/subsys/ipc/openamp/sample.yaml`?

karl-zh commented:
> > I think this PR needs approval before the merge.
> 
> @karl-zh while waiting for approval, could you rebase and fix the conflict with `samples/subsys/ipc/openamp/sample.yaml`?

Hi
 I was trying to resolve the conflict from github, but it seems CI fail. I am on leave for the Chinese New Year holiday recently and will rebase/resolve and try to make CI pass when back to the office. 
The approval and review is required too.

Thanks for your comments.

erwango commented:
@karl-zh, was this intentional to push a merge commit ?
